### PR TITLE
Refactoring to support different map types

### PR DESCRIPTION
**Pull request by DeyuChen**
Refactoring to make developing different map types easier. The refactoring is done using nested controllers. The shared functions are put in the parent controller and the codes for different map types are put in children controllers. The children inherit the $scope namespace from the parent and can access the resource in the parent controller as long as they are under $scope.

### CODE REVIEW

File: examples/twittermap/web/conf/application.conf [hunk 1]
```diff
@@ -117,6 +117,13 @@ tweet.url = "http://twitterMapIndex-search-proxy.herokuapp.com/search/tweets?q=%
 
 # Pointmap parameters
 # How many recent days of the time range predicate of each pointmap query.
-# pointmap.samplingDayRange = 30
+# pointmap.samplingDayRange = 500
```
Comment (waans11) [inline examples/twittermap/web/conf/application.conf]: Why is this change needed?

File: examples/twittermap/web/public/javascripts/common/services.js [hunk 2]
```diff
@@ -47,7 +47,7 @@ angular.module('cloudberry.common', ['cloudberry.mapresultcache'])
     var defaultNonSamplingDayRange = 1500;
     var defaultSamplingDayRange = 1;
     var defaultSamplingSize = 10;
-    var defaultPointmapSamplingDayRange = parseInt(config.pointmapSamplingDayRange);
+    var defaultPointmapSamplingDayRange = 360; //parseInt(config.pointmapSamplingDayRange);
```
Comment (waans11) [inline examples/twittermap/web/public/javascripts/common/services.js]: Why is this change needed?

File: examples/twittermap/web/public/javascripts/common/services.js [hunk 3]
```diff
@@ -473,17 +473,17 @@ angular.module('cloudberry.common', ['cloudberry.mapresultcache'])
             break;
           case "points":
             if(angular.isArray(result.value)) {
-              cloudberryService.tweetResult = result.value[0].slice(0, defaultSamplingSize - 1);
-              cloudberryService.pointsResult = result.value[0];
+              cloudberryService.commonTweetResult = result.value[0].slice(0, defaultSamplingSize - 1);
+              cloudberryService.pointmapMapResult = result.value[0];
```
Comment (waans11) [inline examples/twittermap/web/public/javascripts/common/services.js]: Whenever pointmap is used, let's change it to "pinmap".

File: examples/twittermap/web/public/javascripts/map/heatmap.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
+angular.module('cloudberry.map')
+  .controller('heatMapCtrl', function($scope, $rootScope, $window, $http, $compile, cloudberry, leafletData, cloudberryConfig, Cache) {
+    $rootScope.$on('maptypeChange', function (event, data) {
+      if (cloudberry.parameters.maptype == 'heatmap') {
+        $scope.resetPolygonLayers();
+        //setInfoControlHeatMap();
```
Comment (waans11) [inline examples/twittermap/web/public/javascripts/map/heatmap.js]: Let's remove this commented code.

File: examples/twittermap/web/public/javascripts/map/pointmap.js [hunk 5]
```diff
@@ -0,0 +1,347 @@
+angular.module('cloudberry.map')
+  .controller('pointMapCtrl', function($scope, $rootScope, $window, $http, $compile, cloudberry, leafletData, cloudberryConfig, Cache) {
+    // map type change handler
+    // initialize the map (styles, zoom/drag handler, etc) when switch to this map
+    // clear the map when switch to other map
+    $rootScope.$on('maptypeChange', function (event, data) {
+      if (cloudberry.parameters.maptype == 'pointmap') {
```
Comment (waans11) [inline examples/twittermap/web/public/javascripts/map/pointmap.js]: Let's change it to pinmap. And I think these "countmap", "pinmap", and "heatmap" should be defined as a constant values like Java final string value. We are using these constants again and again.

File: examples/twittermap/web/public/javascripts/map/pointmap.js [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+        setPointMapStyle();
```
Comment (waans11) [inline examples/twittermap/web/public/javascripts/map/pointmap.js]: setPinMapStyle()

File: examples/twittermap/web/public/javascripts/map/pointmap.js [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+        colors: [ '#ffffff', '#92d1e1', '#4393c3', '#2166ac', '#f4a582', '#d6604d', '#b2182b'],
+        sentimentColors: ['#ff0000', '#C0C0C0', '#00ff00']
+      });
+    }
+    
+    // clear pointmap specific data
+    function cleanPointMap() {
+      $scope.points = [];
+      $scope.pointIDs = [];
+      //$scope.map.off('mousemove', $scope.onMapMouseMove);
```
Comment (waans11) [inline examples/twittermap/web/public/javascripts/map/pointmap.js]: Let's remove all the commented codes.

File: examples/twittermap/web/public/javascripts/sidebar/controllers.js [hunk 8]
```diff
@@ -55,14 +55,37 @@ angular.module('cloudberry.sidebar', ['cloudberry.common'])
...
         $scope.result = null;
-        cloudberry.parameters.maptype = 'countmap';
+        cloudberry.parameters.maptype = config.defaultMapType;
 
         var icon1 = document.getElementById('img1');
         var icon2 = document.getElementById('img2');
         var icon3 = document.getElementById('img3');
+        
+        switch (cloudberry.parameters.maptype){
+          case "countmap":
```
Comment (waans11) [inline examples/twittermap/web/public/javascripts/sidebar/controllers.js]: Same here. "countmap", "pinmap", and "heatmap" should be defined as constants.

File: examples/twittermap/web/public/javascripts/map/countmap.js [hunk 9]
```diff
@@ -0,0 +1,476 @@
+angular.module('cloudberry.map')
+  .controller('countMapCtrl', function($scope, $rootScope, $window, $http, $compile, cloudberry, leafletData, cloudberryConfig, Cache) {
+    // map type change handler
+    // initialize the map (styles, zoom/drag handler, etc) when switch to this map
+    // clear the map when switch to other map
+    $rootScope.$on('maptypeChange', function (event, data) {
```
Comment (baiqiushi) [inline examples/twittermap/web/public/javascripts/map/countmap.js]: For this remember old maptype logic, I have some doubts that it looks not clean and elegant.
I think a better way is to let the event data tell the listener this maptype change is from which type changed to which type:
We could modify the sidebar/controller.js to include both pre-maptype and post-maptype into the data of this event.
```
if (cloudberry.parameters.maptype !== 'countmap') {
                var premaptype = cloudberry.parameters.maptype; // * Add one line here * //
                cloudberry.parameters.maptype = 'countmap';
                icon1.src = "/assets/images/aggregation_map.png";
                icon2.src = "/assets/images/heat_map_no_border.png";
                icon3.src = "/assets/images/point_map_no_border.png";
                $rootScope.$emit("maptypeChange", [premaptype, cloudberry.parameters.maptype]); // * Modify this line to give an [] as event data * //
}
```

File: examples/twittermap/web/conf/application.conf (hunk 1 above)
Comment (DeyuChen) [reply examples/twittermap/web/conf/application.conf]: Sorry I change this for testing on my local machine. I will change it back.

File: examples/twittermap/web/public/javascripts/common/services.js (hunk 2 above)
Comment (DeyuChen) [reply examples/twittermap/web/public/javascripts/common/services.js]: I will change it back.

File: examples/twittermap/web/public/javascripts/common/services.js (hunk 3 above)
Comment (DeyuChen) [reply examples/twittermap/web/public/javascripts/common/services.js]: I will do all the renaming in a separate PR.

File: examples/twittermap/web/public/javascripts/map/pointmap.js (hunk 6 above)
Comment (DeyuChen) [reply examples/twittermap/web/public/javascripts/map/pointmap.js]: I will change it in a separate PR

File: examples/twittermap/web/public/javascripts/map/pointmap.js (hunk 7 above)
Comment (DeyuChen) [reply examples/twittermap/web/public/javascripts/map/pointmap.js]: It is removed.

File: examples/twittermap/web/conf/application.conf (hunk 1 above)
[the remark this answers is not in the document]
Comment (DeyuChen) [reply examples/twittermap/web/conf/application.conf]: I have changed it back. Sorry about that.

File: examples/twittermap/web/public/javascripts/common/services.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (DeyuChen) [reply examples/twittermap/web/public/javascripts/common/services.js]: I have change it back.

File: examples/twittermap/web/public/javascripts/map/pointmap.js (hunk 5 above)
Comment (DeyuChen) [reply examples/twittermap/web/public/javascripts/map/pointmap.js]: I am not sure about this. I thought pointmap shouldn't know the existence of other map type. Or maybe I misunderstand what you mean. We should talk about it.
